src/trace_queue.js

```diff
@@ -0,0 +1,188 @@
+                utils.debugLog(`Queue byte size reached ${this.currentByteSize} Bytes, releasing batch... `);
+                this.releaseBatch(this.batchSize - 1);
+            }
+        });
+
+        this.on('batchReleased', (batch) => {
+            this.batchSender(batch);
+            this.emit('batchSent', batch);
+        });
+    }
+
+    /**
+   * Queue size getter
+   * @returns {Number}
+   */
+    get currentSize() {
+        return this.traces.length;
+    }
+
+
+    /**
+   * Checks if queue size reached batch size
+   * @returns {Boolean}
+   */
+    batchSizeReached() {
+        return this.currentSize === this.batchSize;
+    }
+
+    /**
+   * Checks if queue byte size reached its limit
+   * @returns {Boolean}
+   */
+    byteSizeLimitReached() {
+        return this.currentByteSize >= this.batchBytesLimit;
+    }
+
+    /**
+   * add given trace byte size to total byte size
+   * @returns {Number}
+   */
+    addToCurrentByteSize(trace) {
+        this.currentByteSize += JSON.stringify(trace).length;
+    }
+
+    /**
+   * subtract given trace byte size to total byte size
+   * @returns {Number}
+   */
+    subtractFromCurrentByteSize(trace) {
+        this.currentByteSize -= JSON.stringify(trace).length;
+        this.currentByteSize = Math.max(this.currentByteSize, 0);
+    }
+
+    /**
+   * Flush queue
+   */
+    flush() {
+        this.traces = [];
+    }
+
+    /**
+   * Push trace to queue, emit event, and check if queue max queue length reached,
+   * if it does, send batch.
+   * @param {String} trace
+   * @returns {TraceQueue}
+   */
+    push(trace) {
+        const timestamp = new Date();
+        this.traces.push({ trace, timestamp });
+        utils.debugLog(`Trace ${trace} pushed to queue`);
+        this.emit('traceQueued', trace);
+        return this;
+    }
+
+
+    /**
+   * Release batch of traces
+   * @returns {TraceQueue}
+   */
+    releaseBatch(count = this.batchSize) {
+        const batch = [];
+        utils.debugLog(`Releasing batch size ${count}...`);
+        while (batch.length < count && !!this.traces.length) {
+            const shiftedTrace = this.traces.shift().trace;
+            batch.push(shiftedTrace);
+            this.subtractFromCurrentByteSize(shiftedTrace);
+        }
+        this.emit('batchReleased', batch);
+        return this;
+    }
+}
+
+
+/**
+ * Session for the post requests to the collector
+ */
+const session = axios.create({
+    timeout: config.getConfig().sendTimeout,
+    httpAgent: new http.Agent({ keepAlive: true }),
+    httpsAgent: new https.Agent({ keepAlive: true }),
+});
+
+/**
+ * Post given batch to epsagon's infrastructure.
+ * @param {*} batchObject The batch data to send.
+ * @returns {Promise} a promise that is resolved after the batch is posted.
+ */
+function postBatch(batchObject) {
+    utils.debugLog(`Posting batch to ${config.getConfig().traceCollectorURL}...`);
+    utils.debugLog(`Batch: ${JSON.stringify(batchObject, null, 2)}`);
```

Comment:
this is a very expensive debug log, maybe print it only if debug is on (even if it is not printed the call to JSON.stringify is executed)

Reply:
so basically the postBatch() is just a duplicated and customised postTrace() (which is a function already exists in the tracer), I'll definitely remove unnecessary debug logs once we'll get stable. Also took your advice to log only on DEBUG mode.